Load all leases from backend

chaochn47 opened the pull request:
Fix #12535

Reproduce integration test is on https://github.com/chaochn47/etcd/pull/6
Manual e2e test is in https://github.com/etcd-io/etcd/issues/12535#issuecomment-1033222663

@serathius PTAL, thanks!

serathius commented on server/etcdserver/v3_server.go:
```diff
@@ -258,6 +258,9 @@ func (s *EtcdServer) Compact(ctx context.Context, r *pb.CompactionRequest) (*pb.
 }
 
 func (s *EtcdServer) LeaseGrant(ctx context.Context, r *pb.LeaseGrantRequest) (*pb.LeaseGrantResponse, error) {
+	if r.ID < int64(lease.NoLease) {
```
```suggestion
	if r.ID <= int64(lease.NoLease) {
```

leaseID equal 0 represents no lease so it also should not be allowed in API
https://github.com/etcd-io/etcd/blob/170d9b9d73c8358043890332b7f4d7b27dbe7640/server/lease/lessor.go#L35

serathius commented on server/etcdserver/v3_server.go:
```diff
@@ -258,6 +258,9 @@ func (s *EtcdServer) Compact(ctx context.Context, r *pb.CompactionRequest) (*pb.
 }
 
 func (s *EtcdServer) LeaseGrant(ctx context.Context, r *pb.LeaseGrantRequest) (*pb.LeaseGrantResponse, error) {
+	if r.ID < int64(lease.NoLease) {
```
There are other API methods accepting Lease.ID, should add this validation also to other methods?

chaochn47 replied:
Yeah, theoretically other API should return return `LeaseNotFound` if leaseID is negative. But it's good to have the validation in place.

chaochn47 commented:
The test failures are not related to current PR.

Fail at `test_e2e` `TestSerializableReadWithoutQuorum` during the `Coverage/coverage` test while passing the same test in `E2E/test`

```
--- FAIL: TestSerializableReadWithoutQuorum (10.15s)
    logger.go:130: 2022-02-10T07:32:10.353Z	INFO	starting server...	{"name": "test-0"}
    logger.go:130: 2022-02-10T07:32:10.353Z	INFO	spawning process	{"args": ["--name", "test-0", "--listen-client-urls", "http://localhost:20000", "--advertise-client-urls", "http://localhost:20000", "--listen-peer-urls", "https://localhost:20001", "--initial-advertise-peer-urls", "https://localhost:20001", "--initial-cluster-token", "new", "--data-dir", "/tmp/TestSerializableReadWithoutQuorum846452218/001", "--snapshot-count", "100000", "--peer-auto-tls", "--initial-cluster", "test-0=https://localhost:20001,test-1=https://localhost:20006,test-2=https://localhost:20011", "-test.coverprofile=e2e.1644478330353815367.coverprofile", "-test.outputdir=/home/runner/work/etcd/etcd/covdir/e2e"], "working-dir": "/home/runner/work/etcd/etcd/tests/e2e"}
    logger.go:130: 2022-02-10T07:32:15.416Z	INFO	started server.	{"name": "test-0"}
    ctl_v3_test.go:273: ---testFunc logic DONE
    ctl_v3_test.go:286: closing test cluster...
    logger.go:130: 2022-02-10T07:32:15.455Z	INFO	closing test cluster...
    logger.go:130: 2022-02-10T07:32:15.455Z	INFO	stoping server...	{"name": "test-0"}
    logger.go:130: 2022-02-10T07:32:15.483Z	INFO	stopped server.	{"name": "test-0"}
    logger.go:130: 2022-02-10T07:32:15.483Z	INFO	closing server...	{"name": "test-0"}
    logger.go:130: 2022-02-10T07:32:15.483Z	INFO	stoping server...	{"name": "test-0"}
    logger.go:130: 2022-02-10T07:32:15.483Z	INFO	removing directory	{"data-dir": "/tmp/TestSerializableReadWithoutQuorum846452218/001"}
    logger.go:130: 2022-02-10T07:32:15.483Z	INFO	closed test cluster.
    ctl_v3_test.go:289: closed test cluster...
    ctl_v3_kv_no_quorum_test.go:69: linearizableReadTest is expected to fail, but it succeeded
    ctl_v3_test.go:273: ---testFunc logic DONE
    ctl_v3_test.go:286: closing test cluster...
    logger.go:130: 2022-02-10T07:32:20.503Z	INFO	closing test cluster...
    logger.go:130: 2022-02-10T07:32:20.503Z	INFO	stoping server...	{"name": "test-0"}
    logger.go:130: 2022-02-10T07:32:20.503Z	INFO	closing server...	{"name": "test-0"}
    logger.go:130: 2022-02-10T07:32:20.503Z	INFO	stoping server...	{"name": "test-0"}
    logger.go:130: 2022-02-10T07:32:20.503Z	INFO	removing directory	{"data-dir": "/tmp/TestSerializableReadWithoutQuorum846452218/001"}
    logger.go:130: 2022-02-10T07:32:20.503Z	INFO	closed test cluster.
    ctl_v3_test.go:289: closed test cluster...
```

Fail `grpcproxy-integration-test` `TestAuthority` during the `grpcProxy-tests/test` test while passing the same test in `integration-test`. Looks like this specific test is duplicated cause it doesn't explicitly exclude the `grpc-proxy` build tag. 
```
--- FAIL: TestAuthority (2.45s)
    --- FAIL: TestAuthority/Size:_3,_Scenario:_"http://address[:port]" (0.00s)
        logger.go:130: 2022-02-10T07:19:21.859Z	INFO	m-1	LISTEN GRPC	{"member": "m-1", "grpcAddr": "127.0.0.1:33040", "m.Name": "m-1"}
        logger.go:130: 2022-02-10T07:19:21.860Z	INFO	m0	LISTEN GRPC	{"member": "m0", "grpcAddr": "127.0.0.1:33071", "m.Name": "m0"}
        cluster.go:693: listen failed on grpc socket 127.0.0.1:33071 (listen tcp 127.0.0.1:33071: bind: address already in use)
```

serathius commented on server/etcdserver/v3_server.go:
```diff
@@ -271,6 +274,9 @@ func (s *EtcdServer) LeaseGrant(ctx context.Context, r *pb.LeaseGrantRequest) (*
 }
 
 func (s *EtcdServer) LeaseRevoke(ctx context.Context, r *pb.LeaseRevokeRequest) (*pb.LeaseRevokeResponse, error) {
+	if r.ID < int64(lease.NoLease) {
```
```suggestion
	if r.ID <= int64(lease.NoLease) {
```

You should not be able to revoke NoLease

serathius commented on server/etcdserver/v3_server.go:
```diff
@@ -279,6 +285,9 @@ func (s *EtcdServer) LeaseRevoke(ctx context.Context, r *pb.LeaseRevokeRequest)
 }
 
 func (s *EtcdServer) LeaseRenew(ctx context.Context, id lease.LeaseID) (int64, error) {
+	if id < lease.NoLease {
```
```suggestion
	if r.ID <= int64(lease.NoLease) {
```

You should not be able to renew NoLease

serathius commented on server/etcdserver/v3_server.go:
```diff
@@ -314,6 +323,9 @@ func (s *EtcdServer) LeaseRenew(ctx context.Context, id lease.LeaseID) (int64, e
 }
 
 func (s *EtcdServer) LeaseTimeToLive(ctx context.Context, r *pb.LeaseTimeToLiveRequest) (*pb.LeaseTimeToLiveResponse, error) {
+	if r.ID < int64(lease.NoLease) {
```
```suggestion
	if r.ID <= int64(lease.NoLease) {
```

You should not be able to get TTL for NoLease

serathius commented:
I'm still not sure what exactly this will fix. Let's try to describe the current behavior:
* [LeaseGrant](https://github.com/etcd-io/etcd/blob/a905430d27ec7372267b1cf193f6aa6cda68adb6/etcdserver/v3_server.go#L247-L258) accepts 3 ranges of lease ID:
  * Positive - This is a correct ID that will work correctly. 
  * Zero - This will prompt Server to generate it for user, should result in positive ID, https://github.com/etcd-io/etcd/issues/12535#L251
  * Negative - Value allowed in API, however possibly not expected and correctly handled.
* [initAndRecover](https://github.com/etcd-io/etcd/blob/a905430d27ec7372267b1cf193f6aa6cda68adb6/lease/lessor.go#L775) loads only positive leases to memory. So if etcd is restarted lease with negative ID would not be loaded, resulting in lessor resulting returning "lease not found" for it. https://github.com/etcd-io/etcd/blob/a905430d27ec7372267b1cf193f6aa6cda68adb6/lease/lessor.go#L311-L315

Going to the issues reported in https://github.com/etcd-io/etcd/issues/12535
* Etcd state gets corrupted and logs logs like `failed to apply request "header:<ID:9133152155410395282 > lease_revoke:<id:-6f4bd77510edab61>" with response "size:27" took (69.202µs) to execute, err is lease not found`, that share commonality of reporting negative lease ID.

serathius commented:
This leads me to conclude that main problem is related to API allowing negative lease ID, however [initAndRecover](https://github.com/etcd-io/etcd/blob/a905430d27ec7372267b1cf193f6aa6cda68adb6/lease/lessor.go#L775) not handling it correctly. 

However we cannot just disable negative leaseID as this would be a breaking change in the API. I think we should rather consider fixing initAndRecover instead.

serathius commented on tests/integration/v3_lease_test.go:
```diff
@@ -142,26 +143,84 @@ func TestV3LeaseGrantByID(t *testing.T) {
-// TestRejectLeaseRequestWithInvalidID ensures leases request with a negative ID is rejected
-func TestRejectLeaseRequestWithInvalidID(t *testing.T) {
+// TestV3LeaseGrantByNegativeID ensures leases may be created by a negative given id and handled correctly
+func TestV3LeaseGrantByNegativeID(t *testing.T) {
```
```suggestion
func TestIssue12535(t *testing.T) {
```
This is a very specific scenario of testing: add negative lease, restart node, revoke. In this situation I think it's better to refer to issue.

serathius commented on tests/integration/v3_lease_test.go:
```diff
@@ -142,26 +143,84 @@ func TestV3LeaseGrantByID(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	// grant negative leaseID
+	input := []struct {
```
Instead of testing 3 inputs in one cluster and risk side effects (for example if first input breaks revision, error will be reported for all inputs after), could we create subtest for each input that creates it's own cluster?
Don't worry about performance, in tests correctness and clarity matters more.

serathius commented on server/storage/schema/lease.go:
```diff
@@ -27,15 +27,18 @@ func UnsafeCreateLeaseBucket(tx backend.BatchTx) {
 }
 
 func MustUnsafeGetAllLeases(tx backend.ReadTx) []*leasepb.Lease {
```
Could you also add a unit test?

serathius reviewed (approved):
LGTM
cc @ptabor @spzala @ahrtr

spzala commented on tests/integration/v3_lease_test.go:
```diff
@@ -142,6 +143,86 @@ func TestV3LeaseGrantByID(t *testing.T) {
 	}
 }
 
+// TestIssue12535 ensures restarted member can recover negative leaseID from backend and handle it correctly
```
lgtm and thanks @chaochn47 but a small suggestion, can we find a better name for the test? Something like TestV3LeaseNegativeID. The issue title or initial description doesn't give any idea about its relationship with this test unless digging deeper, and in general, always good to have more meaningful name. For easy reference to associated issue, it's fine to put issue number in the doc comment. What do you think? It's okay to change it separately if you prefer (with clean sweep. Thanks!

chaochn47 replied:
Yeah, I will rename the test name and provide more clear test description with comments. 

I need to squash the commits regardless :) @spzala

spzala replied:
:) Awesome, thanks @chaochn47

chaochn47 commented:
Addressed all comments, squashed all commits into one and added one line in `CHANGELOG-3.6.md`. PTAL, thanks!!

I think we should backport this bug fix to `release-3.3`, `release-3.4` and `release-3.5` branches. WDYT? @serathius @spzala

spzala commented:
> Addressed all comments, squashed all commits into one and added one line in `CHANGELOG-3.6.md`. PTAL, thanks!!
> 
> I think we should backport this bug fix to `release-3.3`, `release-3.4` and `release-3.5` branches. WDYT? @serathius @spzala

@chaochn47 thank you for addressing comments and squash, it looks great. +1 to backporting.
